graphix/flow/core.py

```diff
@@ -0,0 +1,488 @@
+            Open graph with respect to which the corrections are defined.
+        x_corrections : Mapping[int, AbstractSet[int]] | None
+            Mapping of X-corrections: in each (`key`, `value`) pair, `key` is a measured node, and `value` is the set of nodes on which an X-correction must be applied depending on the measurement result of `key`.
+        z_corrections : Mapping[int, AbstractSet[int]] | None
+            Mapping of X-corrections: in each (`key`, `value`) pair, `key` is a measured node, and `value` is the set of nodes on which an X-correction must be applied depending on the measurement result of `key`.
+
+        Returns
+        -------
+        XZCorrections[_M_co]
+
+        Notes
+        -----
+        This method computes the partial order induced by the XZ-corrections.
+        """
+        x_corrections = x_corrections or {}
+        z_corrections = z_corrections or {}
+
+        nodes_set = set(og.graph.nodes)
+        outputs_set = frozenset(og.output_nodes)
+        non_outputs_set = nodes_set - outputs_set
+
+        if not non_outputs_set.issuperset(x_corrections):
+            raise ValueError("Keys of input X-corrections contain non-measured nodes.")
+        if not set(z_corrections).issubset(non_outputs_set):
+            raise ValueError("Keys of input Z-corrections contain non-measured nodes.")
+
+        dag = _corrections_to_dag(x_corrections, z_corrections)
+        partial_order_layers = _dag_to_partial_order_layers(dag)
+
+        if partial_order_layers is None:
+            raise ValueError(
+                "Input XZ-corrections are not runnable since the induced directed graph contains closed loops."
+            )
+
+        # If there're no corrections, the partial order has 2 layers only: outputs and measured nodes.
+        if len(partial_order_layers) == 0:
+            partial_order_layers = [outputs_set] if outputs_set else []
+            if non_outputs_set:
+                partial_order_layers.append(frozenset(non_outputs_set))
+            return XZCorrections(og, x_corrections, z_corrections, tuple(partial_order_layers))
+
+        # If the open graph has outputs, the first element in the output of `_dag_to_partial_order_layers(dag)` may or may not contain output nodes.
+        if outputs_set:
+            partial_order_layers = [
+                outputs_set,
+                frozenset(partial_order_layers[0] - outputs_set),
+                *partial_order_layers[1:],
+            ]
+
+        ordered_nodes = frozenset.union(*partial_order_layers)
+
+        if not ordered_nodes.issubset(nodes_set):
+            raise ValueError("Values of input mapping contain labels which are not nodes of the input open graph.")
+
+        # We append to the last layer (first measured nodes) all the non-output nodes not involved in the corrections.
+        if unordered_nodes := frozenset(nodes_set - ordered_nodes):
+            partial_order_layers.append(unordered_nodes)
```

Comment:
Actually, `unordered_nodes` can be added to any layer, they do not need to be put in their own layer. I would say that if `partial_order_layers != []` then `partial_order_layers[-1] = partial_order_layers[-1] | unordered_nodes`.